Tools: add crude calculator for transaction size

harding opened the pull request:
This adds a crude Javascript-based transaction size calculator to `/en/tools/calc-size/`.  No links are added to the tool from other parts of the site and I'm not sure we should add any unless the tool is improved.  

However, I've found it good enough for my own use and I think it might help y'all when you're burdened with verifying my transaction size math in newsletter and workbook content.  Future improvements can let it deal with a wider range of multisig variations, differing types of input and output scripts, non-keypath P2TR, uncompressed pubkeys, low-R signatures, LN onchain scripts, and other common transaction types.

That said, if you don't think this is a good fit for the Optech website, let me know and I can easily host it somewhere else.

jnewbery commented:
I think it's fine to add this. I also agree that we shouldn't link to it until it's a bit more polished.

One thing that I'd like to see added is documentation on how the values are calculated.

harding commented:
> One thing that I'd like to see added is documentation on how the values are calculated.

I don't understand this.  Is this more code documentation (I thought that was ok), more description about what fields are included in a transaction in the HTML output, or something else?

jnewbery commented:
> I don't understand this. Is this more code documentation (I thought that was ok), more description about what fields are included in a transaction in the HTML output, or something else?

More description in the html about how the values are calculated.

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,184 @@
+  document.getElementById('result').innerHTML = (
+    "Total size: " + size + "<br>"
+    + "Overhead: " + overhead + "<br>"
+    + "Input size = " + inputs + " x " + input_size + ' = ' + inputs * input_size + "<br>"
```
I think having the 'vbytes' unit after the sizes would help any ambiguity

bitschmidty replied:
Looks good

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Calculate the result for the default form parameters
+calculateTotal();
```
Id suggest a default input output count here (say 1 in 2 out?) to make initial page load more informative to a new visitor

bitschmidty replied:
Perfecto

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,184 @@
+  var form = document.forms["vbytescalc"];
+
+  // The input/output type
+  // TODO: maybe allow the input type to be different from the output type
```
Also could allow mixed types within inputs and outputs.

Scope creep!

bitschmidty replied:
Future idea, resolving.

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,184 @@
+<form action="" id="vbytescalc" onsubmit="return false;">
+   <fieldset>
+
+    <select id="type" onchange="calculateTotal()">
```
Perhaps "Transaction Type" label for this dropdown field

bitschmidty replied:
Resolving

harding commented:
Forced push, hopefully addressing all comments (including providing more information about how the values are calculated).  Thanks for the reviews, @jnewbery and @bitschmidty!

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+  transaction.  {{page.size.nin}} byte for up to 252
+  inputs
+
+- **Output count** ([compactSize][]) The number of inputs included in
```
s/inputs/outputs

bitschmidty replied:
resolved

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+  of the output (UTXO) being spent
+
+- **scriptSig length** ([compactSize][]) The length of the scriptSig
+  field.  {{page.size.scriptsig_length}} vbytes for a scriptSig up to 252
```
s/vbytes/vbyte (since its 1?)

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+
+- **scriptSig** (varies) The source of witness data for legacy UTXOs.
+  This data is used to prove that the transaction is authorized by
+  someone controlling the appropriate private keys  For the templates
```
s/private keys For/private keys. For

bitschmidty commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+  defined meaning
+
+- **Witness data** (varies) The source of witness data for in segwit
+  transactions.  This data is used to proves that the transaction is
```
s/to proves/to prove

0xB10C commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+
+  <br>
+  Number of inputs: <input type="number" id="inputs" value="1" onchange="calculateTotal()"/><br>
+  Number of outputs: <input type="number" id="outputs" value="2" onchange="calculateTotal()"/>
```
Both HTML form inputs could do with a `min=0` attribute. 

![image](https://user-images.githubusercontent.com/19157360/74068158-d56fab00-49f2-11ea-81f9-e1984a99068e.png)

0xB10C commented on en/tools/calc-size.md:
```diff
@@ -0,0 +1,331 @@
+  // The input/output type
+  // TODO: maybe allow the input type to be different from the output
+  //       type, or multiple inputs of different types (or multiple
+  //       outputs of different types)
```
That would be cool! I might do a follow up PR that adds this.

harding replied:
That'd be great!  Also having someone work on this who actually knows JS would probably be very beneficial.  :-)

harding commented:
Force pushed some edits:

- Hopefully resolved @bitschmidty's editing feedback
- added `min="1"` to input and output count form fields (@0xB10C suggested `min=0`, but since consensus rules require all transactions contain at least one input and one output, I decided to use a minimum of `1`)
- Separated the size calculation logic function from the HTML presentation function
- Added some crude semi-automated testing to verify the form is generating accurate results.  I only tested the default 1-in, 2-out cases; linking each of them (except P2WSH) to an actual transaction on the block chain.  This revealed two minor errors, which I corrected:
    - P2SH 2-of-3 was off by 2 vbytes because its scriptSig field size uses a 3-byte compactSize uint instead of the 1-byte uint I expected
    - P2WSH 2-of-3 was off by 0.25 vbytes because I forgot to include the size of the witnessScript field in the calculation (which would've been obvious if I read my own documentation on the page).  Additionally, using a quick scan of blocks 600,000 to 601,000, I was unable to find a P2WSH 2-of-3 transaction with 1-in, 2-outs that used 72 byte signatures for both of its sigs, so I don't have a txid reference for that.  However, I'm sure it's correct based on a mixed 71/72 byte signature transaction I checked.
- I also disabled breadcrumbs for this page because it wants to link to an `/en/tools/` page that doesn't exist.  Obviously we can add that page later if we want to publicize this

harding commented:
> P2SH 2-of-3 was off by 2 vbytes because its scriptSig field size uses a 3-byte compactSize uint instead of the 1-byte uint I expected

Oh, perhaps worth a dorky note, I think that if at least one signature for P2SH 2-of-3 uses low-R (and all pubkeys are compressed), then this can be a 1-byte compactSize uint.  Obviously, anyone who cares about space savings should be using P2WSH instead, but it might be worth a footnote in the scaling workbook or somewhere else that using low-R sigs can save you up to 4 vbytes here instead of just the expected 2 vbytes.  🤓

bitschmidty commented:
tACK 6d0712c57afdbc68cf209dc22c790be8f626d709 (Chrome, Safari, Firefox on macOS)

harding commented:
Pushed a single line edit to make the P2WSH 2-of-3 witness example more accurate and less confusing.

jnewbery commented:
Looks great. Thanks Dave!
